### PR TITLE
Implement SignAndSendTransaction, SignTransaction, and SignAllTransactions in Solana provider

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/21766

This PR only implements the backend part, currently we don't have front-end codes showing and processing SignTransaction and SignAllTransactions. And as for SignAndSendTransaction, even if users enable Solana dapp feature flag manually, since we haven't update the UI screen to process it correctly, users won't be able to confirm it either, the screen currently looks as below.
Screen Shot 2022-05-27 at 3 57 33 PM

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/solana_tx_manager.cc [hunk 1]
```diff
@@ -204,7 +210,8 @@ void SolanaTxManager::OnGetSignatureStatuses(
       continue;
 
     if (!signature_statuses[i]) {
-      if (meta->tx()->message()->last_valid_block_height() < block_height) {
+      if (meta->tx()->message()->last_valid_block_height() &&
```
Comment (yrliou) [inline components/brave_wallet/browser/solana_tx_manager.cc:213]: For transactions created from dapps, we can't get this info so last_valid_block_height will be kept as 0 (the initialized value).
Created a follow-up issue to support dapp transactions by calling IsBlockhashValid at https://github.com/brave/brave-browser/issues/23120.

**Comment by yrliou** [comment]
Optional SendOptions param for signAndSendTransaction is not supported by this PR yet, will address in https://github.com/brave/brave-browser/issues/23122.

**Review by darkdh — APPROVED** [review]
++ just one nit

File: components/brave_wallet/browser/solana_provider_impl.cc [hunk 2]
```diff
@@ -114,6 +122,37 @@ void SolanaProviderImpl::GetPublicKey(GetPublicKeyCallback callback) {
...
+  // 1. Serialize the created message object and encode in Base58 will be the
+  //    same as original encoded message.
+  // 2. Fee payer should be the current selected account.
+  // 3. Only one signer (the fee payer) is allowed, we do not support multisig
+  //    for dApp requests currently.
+  std::vector<std::string> signers;
+  auto serialized_message = msg->Serialize(&signers);
+  if (!serialized_message ||
+      Base58Encode(*serialized_message) != encoded_serialized_msg ||
+      account != msg->fee_payer() || signers.size() != 1) {
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/solana_provider_impl.cc]: Updated here https://github.com/brave/brave-core/pull/13439/commits/00eced812d8a793ea17b9be74d667e82c97eb247#diff-072a34f5cbcd051507b00f9df51f345c6c4ab01d4381cb90ba3820708cf5e82eR139-R152

**Review by kylehickinson — APPROVED** [review]
iOS++

**Comment by yrliou** [comment]
Windows x86 failed at BraveSchemeLoadBrowserTest.CrashURLTest which is known and unrelated, merging.

**Pull request merged**